### PR TITLE
Add Chafa, a library for images in the terminal

### PR DESCRIPTION
**Pull request by ararslan**
Let's see whether this works.

### CODE REVIEW

**Comment by ararslan** [comment]
Doesn't

File: C/Chafa/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+version = v"1.4.1"
+
+sources = [
+    ArchiveSource("https://hpjansson.org/chafa/releases/chafa-$(version).tar.xz",
+                  "46d34034f4c96d120e0639f87a26590427cc29e95fe5489e903a48ec96402ba3"),
+]
+
+script = raw"""
+cd ${WORKSPACE}/srcdir/chafa-*/
+./autogen.sh
```
Comment (giordano) [inline C/Chafa/build_tarballs.jl:13]:
```suggestion
./autogen.sh --prefix=$prefix --build=${MACHTYPE} --host=${target}
```
I guess?

Comment (ararslan) [reply C/Chafa/build_tarballs.jl]: Ah, I got the prefix and host but not `MACHTYPE`; what is that?

Comment (giordano) [reply C/Chafa/build_tarballs.jl]: `${MACHTYPE}` is the triplet of the build environment, so `x86_64-linux-musl` everywhere.  Sometimes build systems explicitly ask for it

**Comment by giordano** [comment]
Given what are the four platforms that are building correctly, without opening the log file I can guess the build system is trying to run a program built for the target platform

File: C/Chafa/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+version = v"1.4.1"
+
+sources = [
+    ArchiveSource("https://hpjansson.org/chafa/releases/chafa-$(version).tar.xz",
+                  "46d34034f4c96d120e0639f87a26590427cc29e95fe5489e903a48ec96402ba3"),
+]
+
+script = raw"""
+cd ${WORKSPACE}/srcdir/chafa-*/
+./autogen.sh --prefix=${prefix} --host=${target}
```
Comment (ararslan) [inline C/Chafa/build_tarballs.jl:13]:
```suggestion
./autogen.sh --prefix=${prefix} --build=${MACHTYPE} --host=${target}
```

**Comment by ararslan** [comment]
> [17:08:35] configure: error: cannot run test program while cross compiling

So don't do that, then :weary:

**Comment by giordano** [comment]
You should look into what's the program that they're trying to run and why.  Maybe we can cache the value of the test without actually performing it

**Comment by giordano** [comment]
For the record, it's trying to build and run the following program:
```c
#include <stdlib.h>
int
main ()
{
__builtin_cpu_init ();
  ;
  return 0;
}
```
The result can be cached

**Comment by ararslan** [comment]
Yeah I saw that yesterday and started looking into it briefly but didn't determine whether that was an important check that we could disable/ignore/cache. How do we cache it?

**Comment by giordano** [comment]
By calling `./autogen.sh` with
```
./autogen.sh ... ax_cv_gcc_check_x86_cpu_init=yes ax_cv_gcc_check_x86_cpu_supports=yes
```

File: C/Chafa/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+version = v"1.4.1"
+
+sources = [
+    ArchiveSource("https://hpjansson.org/chafa/releases/chafa-$(version).tar.xz",
+                  "46d34034f4c96d120e0639f87a26590427cc29e95fe5489e903a48ec96402ba3"),
+]
+
+script = raw"""
+cd ${WORKSPACE}/srcdir/chafa-*/
+./autogen.sh --prefix=${prefix} --build=${MACHTYPE} --host=${target}
```
Comment (ararslan) [inline C/Chafa/build_tarballs.jl:13]:
```suggestion
./autogen.sh \
    --prefix=${prefix} \
    --build=${MACHTYPE} \
    --host=${target} \
    ax_cv_gcc_check_x86_cpu_init=yes \
    ax_cv_gcc_check_x86_cpu_supports=yes
```

Comment (giordano) [reply C/Chafa/build_tarballs.jl]: Ok, now we've got FreeBSD in, I guess that's enough for you? :stuck_out_tongue:   

Jokes apart, my understanding is that this is true only when `${proc_family} == intel`

**Comment by ararslan** [comment]
Caching that fixed i686 and x86-64 Linux as well as FreeBSD, but other things aren't happy. This error from the macOS build seems troubling:
> ld: illegal text reloc in '_init_once' to '___cpu_model' for architecture x86_64

**Comment by giordano** [comment]
Yeah, I don't see anything special in the [Homebrew recipe](https://github.com/Homebrew/homebrew-core/blob/2d95073b0ae557fcd1f4a0267be22be388876173/Formula/chafa.rb) for macOS.

**Comment by giordano** [comment]
macOS works for me by building with GCC:
```sh
export CC=gcc
export CXX=g++
```
before running `./autogen.sh`.  I'm not sure Windows is supposed to work at all, it's not mentioned here: https://hpjansson.org/chafa/download/.  Also the non-intel platforms may not be expected to work

**Comment by ararslan** [comment]
It works for you locally with GCC or in the BinaryBuilder environment? Looks like the Homebrew formula doesn't require building with GCC, so it should at least theoretically be able to be built with Clang... :thinking: Could be another case of software that doesn't like to be cross compiled, like HDF5.

**Comment by giordano** [comment]
> It works for you locally with GCC or in the BinaryBuilder environment?

In BB

**Comment by giordano** [comment]
After grepping some keywords in the code, I'm more and more convinced that this doesn't support Windows

File: C/Chafa/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    --prefix=${prefix} \
+    --build=${MACHTYPE} \
+    --host=${target} \
+    ax_cv_gcc_check_x86_cpu_init="${BUILTIN_FUNCS}" \
+    ax_cv_gcc_check_x86_cpu_supports="${BUILTIN_FUNCS}"
+make -j${nproc}
+make install
+"""
+
+platforms = supported_platforms()
```
Comment (ararslan) [inline C/Chafa/build_tarballs.jl:29]:
```suggestion
# Chafa itself does not support Windows
platforms = filter!(p -> !isa(p, Windows), supported_platforms())
```

File: C/Chafa/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+version = v"1.4.1"
+
+sources = [
+    ArchiveSource("https://hpjansson.org/chafa/releases/chafa-$(version).tar.xz",
+                  "46d34034f4c96d120e0639f87a26590427cc29e95fe5489e903a48ec96402ba3"),
+]
+
+script = raw"""
+cd ${WORKSPACE}/srcdir/chafa-*/
+
```
Comment (ararslan) [inline C/Chafa/build_tarballs.jl:13]:
```suggestion

if [[ ${target} == *darwin* ]]; then
    # For some reason building with Clang for macOS doesn't work
    export CC=gcc
fi
```

Comment (ararslan) [reply C/Chafa/build_tarballs.jl:13]: Is this what you had in mind, @giordano?

Comment (giordano) [reply C/Chafa/build_tarballs.jl:13]: Yes, I think the build system is quite messy, the use of `AC_TRY_RUN` to detect the availability of functions in the compiler speaks for itself.  Maybe GCC here works just by chance, but there isn't a fundamental reason why Clang shouldn't work

**Comment by giordano** [comment]
After several attempts, I managed to fix registration, this is the PR for the registry: https://github.com/JuliaRegistries/General/pull/15622

**Pull request merged**